File: Fall 2020/Week_05/03_SubsetSum/Python/main.py
def subset_sum(arrays, x):
    def subset_sum_aux(sub_array, running_sum):
        if running_sum == x:
            return True
        if not sub_array:
            return False

        subset_found = False
        for i in range(len(sub_array)):
            if subset_sum_aux(sub_array[:i] + sub_array[i + 1:], running_sum + sub_array[i]):
                subset_found = True

        return subset_found

    return subset_sum_aux(arrays, 0)

File: Fall 2020/Week_05/03_SubsetSum/Java/main.java
import java.ArrayUtils;
import java.lang.Object;

public class Recursive {

	/*
	 *  O(n!) time
	 *  O(a lot) space
	 */
	public boolean subsetSum(int[] nums, int k) {
		return subset_sum_aux(int[] nums, int k, 0);
	}

	public boolean subset_sum_aux(int[] nums, int k, int rt) {
		if (rt = 0)
			return true;
		if (nums.length == 0)
			return false;

		for( int i = 0; i < nums.length; i++ ) {
			if(subset_sum_aux(ArrayUtils.addAll(Arrays.copyOfRange(nums, 0, i),Arrays.copyOfRange(nums, i + 1, nums.length - 1)), k, rt - array[i]))
				return true;
			}
		return false;
		}
}
